TestWatchCancelRunning fails

```
=== RUN   TestWatchCancelRunning
2016-06-02 05:01:14.664107 I | integration: launching node3421125399296241962 (unix://localhost:node3421125399296241962.sock)
2016-06-02 05:01:14.664790 I | integration: launching node7419267789755234383 (unix://localhost:node7419267789755234383.sock)
2016-06-02 05:01:14.665615 I | integration: launching node1829180572306803275 (unix://localhost:node1829180572306803275.sock)
2016-06-02 05:01:14.684924 I | etcdserver: name = node3421125399296241962
2016-06-02 05:01:14.684978 I | etcdserver: data dir = /tmp/etcd384540070
2016-06-02 05:01:14.685020 I | etcdserver: member dir = /tmp/etcd384540070/member
2016-06-02 05:01:14.685052 I | etcdserver: heartbeat = 10ms
2016-06-02 05:01:14.685078 I | etcdserver: election = 100ms
2016-06-02 05:01:14.685103 I | etcdserver: snapshot count = 0
2016-06-02 05:01:14.685138 I | etcdserver: advertise client URLs = http://127.0.0.1:21398
2016-06-02 05:01:14.685169 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21397
2016-06-02 05:01:14.685226 I | etcdserver: initial cluster = node1829180572306803275=http://127.0.0.1:21401,node3421125399296241962=http://127.0.0.1:21397,node7419267789755234383=http://127.0.0.1:21399
2016-06-02 05:01:14.685705 I | etcdserver: name = node7419267789755234383
2016-06-02 05:01:14.685753 I | etcdserver: data dir = /tmp/etcd101299917
2016-06-02 05:01:14.685793 I | etcdserver: member dir = /tmp/etcd101299917/member
2016-06-02 05:01:14.685822 I | etcdserver: heartbeat = 10ms
2016-06-02 05:01:14.685848 I | etcdserver: election = 100ms
2016-06-02 05:01:14.685873 I | etcdserver: snapshot count = 0
2016-06-02 05:01:14.685906 I | etcdserver: advertise client URLs = http://127.0.0.1:21400
2016-06-02 05:01:14.685939 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21399
2016-06-02 05:01:14.685994 I | etcdserver: initial cluster = node1829180572306803275=http://127.0.0.1:21401,node3421125399296241962=http://127.0.0.1:21397,node7419267789755234383=http://127.0.0.1:21399
2016-06-02 05:01:14.687166 I | etcdserver: name = node1829180572306803275
2016-06-02 05:01:14.687247 I | etcdserver: data dir = /tmp/etcd180210120
2016-06-02 05:01:14.687308 I | etcdserver: member dir = /tmp/etcd180210120/member
2016-06-02 05:01:14.687357 I | etcdserver: heartbeat = 10ms
2016-06-02 05:01:14.687404 I | etcdserver: election = 100ms
2016-06-02 05:01:14.687446 I | etcdserver: snapshot count = 0
2016-06-02 05:01:14.687505 I | etcdserver: advertise client URLs = http://127.0.0.1:21402
2016-06-02 05:01:14.687565 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21401
2016-06-02 05:01:14.687662 I | etcdserver: initial cluster = node1829180572306803275=http://127.0.0.1:21401,node3421125399296241962=http://127.0.0.1:21397,node7419267789755234383=http://127.0.0.1:21399
2016-06-02 05:01:14.706305 I | etcdserver: starting member ba04f747a29d99e0 in cluster b797a30b04501f03
2016-06-02 05:01:14.706554 I | raft: ba04f747a29d99e0 became follower at term 0
2016-06-02 05:01:14.706644 I | raft: newRaft ba04f747a29d99e0 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-06-02 05:01:14.706699 I | raft: ba04f747a29d99e0 became follower at term 1
2016-06-02 05:01:14.707941 I | etcdserver: starting member d4f4faaba288b790 in cluster b797a30b04501f03
2016-06-02 05:01:14.708173 I | raft: d4f4faaba288b790 became follower at term 0
2016-06-02 05:01:14.708242 I | raft: newRaft d4f4faaba288b790 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-06-02 05:01:14.708285 I | raft: d4f4faaba288b790 became follower at term 1
2016-06-02 05:01:14.708539 I | rafthttp: starting peer 36b7d9d75d3a1f98...
2016-06-02 05:01:14.708822 I | rafthttp: started HTTP pipelining with peer 36b7d9d75d3a1f98
2016-06-02 05:01:14.713803 I | etcdserver: starting member 36b7d9d75d3a1f98 in cluster b797a30b04501f03
2016-06-02 05:01:14.713988 I | raft: 36b7d9d75d3a1f98 became follower at term 0
2016-06-02 05:01:14.714065 I | raft: newRaft 36b7d9d75d3a1f98 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-06-02 05:01:14.714110 I | raft: 36b7d9d75d3a1f98 became follower at term 1
2016-06-02 05:01:14.717998 I | rafthttp: starting peer 36b7d9d75d3a1f98...
2016-06-02 05:01:14.718982 I | rafthttp: started HTTP pipelining with peer 36b7d9d75d3a1f98
2016-06-02 05:01:14.720402 I | rafthttp: started peer 36b7d9d75d3a1f98
2016-06-02 05:01:14.720705 I | rafthttp: added peer 36b7d9d75d3a1f98
2016-06-02 05:01:14.720820 I | rafthttp: starting peer d4f4faaba288b790...
2016-06-02 05:01:14.721170 I | rafthttp: started HTTP pipelining with peer d4f4faaba288b790
2016-06-02 05:01:14.725072 I | rafthttp: started peer 36b7d9d75d3a1f98
2016-06-02 05:01:14.725303 I | rafthttp: added peer 36b7d9d75d3a1f98
2016-06-02 05:01:14.725375 I | rafthttp: starting peer ba04f747a29d99e0...
2016-06-02 05:01:14.725636 I | rafthttp: started HTTP pipelining with peer ba04f747a29d99e0
2016-06-02 05:01:14.726413 I | rafthttp: started streaming with peer d4f4faaba288b790 (writer)
2016-06-02 05:01:14.731732 I | rafthttp: started streaming with peer 36b7d9d75d3a1f98 (stream Message reader)
2016-06-02 05:01:14.739775 I | rafthttp: started streaming with peer 36b7d9d75d3a1f98 (writer)
2016-06-02 05:01:14.739854 I | rafthttp: started streaming with peer 36b7d9d75d3a1f98 (writer)
2016-06-02 05:01:14.739947 I | rafthttp: started streaming with peer 36b7d9d75d3a1f98 (stream MsgApp v2 reader)
2016-06-02 05:01:14.740227 I | rafthttp: started streaming with peer 36b7d9d75d3a1f98 (stream Message reader)
2016-06-02 05:01:14.740375 I | rafthttp: started peer d4f4faaba288b790
2016-06-02 05:01:14.740529 I | rafthttp: added peer d4f4faaba288b790
2016-06-02 05:01:14.740575 I | etcdserver: set snapshot count to default 10000
2016-06-02 05:01:14.740619 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-06-02 05:01:14.740652 I | rafthttp: started streaming with peer d4f4faaba288b790 (writer)
2016-06-02 05:01:14.740708 I | rafthttp: started streaming with peer 36b7d9d75d3a1f98 (writer)
2016-06-02 05:01:14.740774 I | rafthttp: started streaming with peer 36b7d9d75d3a1f98 (writer)
2016-06-02 05:01:14.743988 I | rafthttp: started peer ba04f747a29d99e0
2016-06-02 05:01:14.744148 I | rafthttp: added peer ba04f747a29d99e0
2016-06-02 05:01:14.744186 I | etcdserver: set snapshot count to default 10000
2016-06-02 05:01:14.744217 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-06-02 05:01:14.745139 I | rafthttp: started streaming with peer ba04f747a29d99e0 (writer)
2016-06-02 05:01:14.745224 I | rafthttp: started streaming with peer ba04f747a29d99e0 (stream MsgApp v2 reader)
2016-06-02 05:01:14.745410 I | rafthttp: started streaming with peer ba04f747a29d99e0 (stream Message reader)
2016-06-02 05:01:14.747610 I | membership: added member 36b7d9d75d3a1f98 [http://127.0.0.1:21401] to cluster b797a30b04501f03
2016-06-02 05:01:14.747969 I | membership: added member ba04f747a29d99e0 [http://127.0.0.1:21397] to cluster b797a30b04501f03
2016-06-02 05:01:14.748483 I | membership: added member d4f4faaba288b790 [http://127.0.0.1:21399] to cluster b797a30b04501f03
2016-06-02 05:01:14.750601 I | rafthttp: starting peer ba04f747a29d99e0...
2016-06-02 05:01:14.750914 I | rafthttp: started HTTP pipelining with peer ba04f747a29d99e0
2016-06-02 05:01:14.756111 I | rafthttp: started peer ba04f747a29d99e0
2016-06-02 05:01:14.756318 I | rafthttp: added peer ba04f747a29d99e0
2016-06-02 05:01:14.756379 I | rafthttp: starting peer d4f4faaba288b790...
2016-06-02 05:01:14.756566 I | rafthttp: started streaming with peer d4f4faaba288b790 (stream Message reader)
2016-06-02 05:01:14.756659 I | rafthttp: started HTTP pipelining with peer d4f4faaba288b790
2016-06-02 05:01:14.759892 I | rafthttp: peer d4f4faaba288b790 became active
2016-06-02 05:01:14.759963 I | rafthttp: established a TCP streaming connection with peer d4f4faaba288b790 (stream Message writer)
2016-06-02 05:01:14.760382 I | rafthttp: started peer d4f4faaba288b790
2016-06-02 05:01:14.760584 I | rafthttp: added peer d4f4faaba288b790
2016-06-02 05:01:14.760631 I | etcdserver: set snapshot count to default 10000
2016-06-02 05:01:14.760670 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-06-02 05:01:14.760809 I | rafthttp: established a TCP streaming connection with peer d4f4faaba288b790 (stream MsgApp v2 writer)
2016-06-02 05:01:14.761566 I | rafthttp: started streaming with peer 36b7d9d75d3a1f98 (stream MsgApp v2 reader)
2016-06-02 05:01:14.762108 I | rafthttp: started streaming with peer d4f4faaba288b790 (stream MsgApp v2 reader)
2016-06-02 05:01:14.762365 I | membership: added member 36b7d9d75d3a1f98 [http://127.0.0.1:21401] to cluster b797a30b04501f03
2016-06-02 05:01:14.762596 I | rafthttp: started streaming with peer ba04f747a29d99e0 (writer)
2016-06-02 05:01:14.762679 I | rafthttp: started streaming with peer ba04f747a29d99e0 (writer)
2016-06-02 05:01:14.762790 I | rafthttp: started streaming with peer ba04f747a29d99e0 (stream MsgApp v2 reader)
2016-06-02 05:01:14.763071 I | rafthttp: started streaming with peer ba04f747a29d99e0 (stream Message reader)
2016-06-02 05:01:14.763408 I | rafthttp: started streaming with peer d4f4faaba288b790 (writer)
2016-06-02 05:01:14.763483 I | rafthttp: started streaming with peer d4f4faaba288b790 (writer)
2016-06-02 05:01:14.763538 I | membership: added member ba04f747a29d99e0 [http://127.0.0.1:21397] to cluster b797a30b04501f03
2016-06-02 05:01:14.763584 I | rafthttp: started streaming with peer d4f4faaba288b790 (stream MsgApp v2 reader)
2016-06-02 05:01:14.763930 I | rafthttp: started streaming with peer d4f4faaba288b790 (stream Message reader)
2016-06-02 05:01:14.764932 I | membership: added member d4f4faaba288b790 [http://127.0.0.1:21399] to cluster b797a30b04501f03
2016-06-02 05:01:14.765539 I | rafthttp: started streaming with peer ba04f747a29d99e0 (writer)
2016-06-02 05:01:14.765595 I | integration: launched node7419267789755234383 (unix://localhost:node7419267789755234383.sock)
2016-06-02 05:01:14.765673 I | integration: launched node3421125399296241962 (unix://localhost:node3421125399296241962.sock)
2016-06-02 05:01:14.767541 I | membership: added member 36b7d9d75d3a1f98 [http://127.0.0.1:21401] to cluster b797a30b04501f03
2016-06-02 05:01:14.767876 I | integration: launched node1829180572306803275 (unix://localhost:node1829180572306803275.sock)
2016-06-02 05:01:14.769506 I | rafthttp: peer ba04f747a29d99e0 became active
2016-06-02 05:01:14.769576 I | rafthttp: established a TCP streaming connection with peer ba04f747a29d99e0 (stream Message reader)
2016-06-02 05:01:14.770002 I | rafthttp: peer ba04f747a29d99e0 became active
2016-06-02 05:01:14.770074 I | rafthttp: established a TCP streaming connection with peer ba04f747a29d99e0 (stream MsgApp v2 writer)
2016-06-02 05:01:14.770143 I | rafthttp: peer d4f4faaba288b790 became active
2016-06-02 05:01:14.770203 I | rafthttp: established a TCP streaming connection with peer d4f4faaba288b790 (stream Message writer)
2016-06-02 05:01:14.770806 I | raft: ba04f747a29d99e0 is starting a new election at term 1
2016-06-02 05:01:14.770893 I | raft: ba04f747a29d99e0 became candidate at term 2
2016-06-02 05:01:14.770943 I | raft: ba04f747a29d99e0 received vote from ba04f747a29d99e0 at term 2
2016-06-02 05:01:14.771002 I | raft: ba04f747a29d99e0 [logterm: 1, index: 3] sent vote request to 36b7d9d75d3a1f98 at term 2
2016-06-02 05:01:14.771066 I | raft: ba04f747a29d99e0 [logterm: 1, index: 3] sent vote request to d4f4faaba288b790 at term 2
2016-06-02 05:01:14.771725 I | rafthttp: established a TCP streaming connection with peer ba04f747a29d99e0 (stream MsgApp v2 writer)
2016-06-02 05:01:14.771787 I | rafthttp: established a TCP streaming connection with peer ba04f747a29d99e0 (stream Message writer)
2016-06-02 05:01:14.772570 I | rafthttp: peer 36b7d9d75d3a1f98 became active
2016-06-02 05:01:14.772619 I | membership: added member ba04f747a29d99e0 [http://127.0.0.1:21397] to cluster b797a30b04501f03
2016-06-02 05:01:14.772762 I | rafthttp: established a TCP streaming connection with peer ba04f747a29d99e0 (stream MsgApp v2 reader)
2016-06-02 05:01:14.772996 I | rafthttp: established a TCP streaming connection with peer 36b7d9d75d3a1f98 (stream MsgApp v2 writer)
2016-06-02 05:01:14.774254 I | membership: added member d4f4faaba288b790 [http://127.0.0.1:21399] to cluster b797a30b04501f03
2016-06-02 05:01:14.775153 I | rafthttp: established a TCP streaming connection with peer d4f4faaba288b790 (stream MsgApp v2 reader)
2016-06-02 05:01:14.775967 I | rafthttp: established a TCP streaming connection with peer d4f4faaba288b790 (stream MsgApp v2 reader)
2016-06-02 05:01:14.776198 I | raft: 36b7d9d75d3a1f98 [term: 1] received a MsgVote message with higher term from ba04f747a29d99e0 [term: 2]
2016-06-02 05:01:14.776314 I | raft: 36b7d9d75d3a1f98 became follower at term 2
2016-06-02 05:01:14.776369 I | raft: 36b7d9d75d3a1f98 [logterm: 1, index: 3, vote: 0] voted for ba04f747a29d99e0 [logterm: 1, index: 3] at term 2
2016-06-02 05:01:14.777029 I | rafthttp: established a TCP streaming connection with peer 36b7d9d75d3a1f98 (stream Message reader)
2016-06-02 05:01:14.777089 I | rafthttp: peer 36b7d9d75d3a1f98 became active
2016-06-02 05:01:14.777145 I | rafthttp: established a TCP streaming connection with peer 36b7d9d75d3a1f98 (stream MsgApp v2 reader)
2016-06-02 05:01:14.777647 I | rafthttp: established a TCP streaming connection with peer ba04f747a29d99e0 (stream Message writer)
2016-06-02 05:01:14.777908 I | raft: d4f4faaba288b790 [term: 1] received a MsgVote message with higher term from ba04f747a29d99e0 [term: 2]
2016-06-02 05:01:14.777979 I | rafthttp: established a TCP streaming connection with peer 36b7d9d75d3a1f98 (stream Message reader)
2016-06-02 05:01:14.778013 I | raft: d4f4faaba288b790 became follower at term 2
2016-06-02 05:01:14.778075 I | raft: d4f4faaba288b790 [logterm: 1, index: 3, vote: 0] voted for ba04f747a29d99e0 [logterm: 1, index: 3] at term 2
2016-06-02 05:01:14.778381 I | raft: ba04f747a29d99e0 received vote from 36b7d9d75d3a1f98 at term 2
2016-06-02 05:01:14.778428 I | raft: ba04f747a29d99e0 [quorum:2] has received 2 votes and 0 vote rejections
2016-06-02 05:01:14.778485 I | raft: ba04f747a29d99e0 became leader at term 2
2016-06-02 05:01:14.778528 I | raft: raft.node: ba04f747a29d99e0 elected leader ba04f747a29d99e0 at term 2
2016-06-02 05:01:14.779582 I | rafthttp: established a TCP streaming connection with peer d4f4faaba288b790 (stream Message reader)
2016-06-02 05:01:14.795520 I | rafthttp: established a TCP streaming connection with peer 36b7d9d75d3a1f98 (stream Message writer)
2016-06-02 05:01:14.800295 I | raft: raft.node: 36b7d9d75d3a1f98 elected leader ba04f747a29d99e0 at term 2
2016-06-02 05:01:14.800674 I | rafthttp: established a TCP streaming connection with peer ba04f747a29d99e0 (stream MsgApp v2 reader)
2016-06-02 05:01:14.800955 I | rafthttp: established a TCP streaming connection with peer 36b7d9d75d3a1f98 (stream MsgApp v2 writer)
2016-06-02 05:01:14.801281 I | raft: raft.node: d4f4faaba288b790 elected leader ba04f747a29d99e0 at term 2
2016-06-02 05:01:14.801650 I | rafthttp: established a TCP streaming connection with peer 36b7d9d75d3a1f98 (stream MsgApp v2 reader)
2016-06-02 05:01:14.802561 I | rafthttp: established a TCP streaming connection with peer 36b7d9d75d3a1f98 (stream Message writer)
2016-06-02 05:01:14.802623 I | rafthttp: established a TCP streaming connection with peer d4f4faaba288b790 (stream MsgApp v2 writer)
2016-06-02 05:01:14.802764 I | rafthttp: established a TCP streaming connection with peer d4f4faaba288b790 (stream Message reader)
2016-06-02 05:01:14.803126 I | rafthttp: established a TCP streaming connection with peer ba04f747a29d99e0 (stream Message reader)
2016-06-02 05:01:14.804987 I | etcdserver: published {Name:node1829180572306803275 ClientURLs:[http://127.0.0.1:21402]} to cluster b797a30b04501f03
2016-06-02 05:01:14.805516 I | etcdserver: published {Name:node3421125399296241962 ClientURLs:[http://127.0.0.1:21398]} to cluster b797a30b04501f03
2016-06-02 05:01:14.808099 I | etcdserver: published {Name:node7419267789755234383 ClientURLs:[http://127.0.0.1:21400]} to cluster b797a30b04501f03
2016-06-02 05:01:14.808645 I | etcdserver: setting up the initial cluster version to 3.0
2016-06-02 05:01:14.809840 N | membership: set the initial cluster version to 3.0
2016-06-02 05:01:14.809973 N | membership: set the initial cluster version to 3.0
2016-06-02 05:01:14.810915 N | membership: set the initial cluster version to 3.0
panic: watcher not victim but not in watch groups

goroutine 28429 [running]:
panic(0xf923c0, 0xc8221368b0)
	/usr/local/golang/1.6.2/go/src/runtime/panic.go:481 +0x3ff
github.com/coreos/etcd/mvcc.(*watchableStore).cancelWatcher(0xc825aed290, 0xc825ba45f0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/mvcc/watchable_store.go:235 +0x1db
github.com/coreos/etcd/mvcc.(*watchableStore).watch.func1()
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/mvcc/watchable_store.go:217 +0x37
github.com/coreos/etcd/mvcc.(*watchStream).Cancel(0xc822756780, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/mvcc/watcher.go:124 +0xdc
github.com/coreos/etcd/etcdserver/api/v3rpc.(*serverWatchStream).recvLoop(0xc8227eb960, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:190 +0x340
github.com/coreos/etcd/etcdserver/api/v3rpc.(*watchServer).Watch.func2(0xc825afb260, 0xc8227eb960)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:126 +0x2f
created by github.com/coreos/etcd/etcdserver/api/v3rpc.(*watchServer).Watch
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:126 +0x2fc
exit status 2
```

--- a/mvcc/watcher.go
+++ b/mvcc/watcher.go
@@ -117,13 +117,18 @@ func (ws *watchStream) Chan() <-chan WatchResponse {
 }
 
 func (ws *watchStream) Cancel(id WatchID) error {
+	ws.mu.Lock()
 	cancel, ok := ws.cancels[id]
+	ok = ok && !ws.closed
+	if ok {
+		delete(ws.cancels, id)
+		delete(ws.watchers, id)
+	}
+	ws.mu.Unlock()
 	if !ok {
 		return ErrWatcherNotExist
 	}
 	cancel()
-	delete(ws.cancels, id)
-	delete(ws.watchers, id)
 	return nil
 }
 

